runtime/shared_common/CacheMap.cpp

```diff
@@ -883,39 +883,29 @@ SH_CacheMap::createNewSegment(J9VMThread* currentThread, UDATA type, J9MemorySeg
 void
 SH_CacheMap::updateROMSegmentList(J9VMThread* currentThread, bool hasClassSegmentMutex)
 {
-	SH_CompositeCacheImpl* cache;
+	SH_CompositeCacheImpl* cache = _ccHead;
 #if defined(J9VM_THR_PREEMPTIVE)
-	J9JavaVM* vm = currentThread->javaVM;
-
-	omrthread_monitor_t classSegmentMutex = vm->classMemorySegments->segmentMutex;
+	omrthread_monitor_t classSegmentMutex = currentThread->javaVM->classMemorySegments->segmentMutex;
 #endif
 	
 #if defined(J9VM_THR_PREEMPTIVE)
-	if (classSegmentMutex) {
-		if (!hasClassSegmentMutex) {
-			Trc_SHR_Assert_ShouldNotHaveLocalMutex(classSegmentMutex);
-			enterLocalMutex(currentThread, classSegmentMutex, "class segment mutex", "updateROMSegmentList");
-		} else {
-			Trc_SHR_Assert_ShouldHaveLocalMutex(classSegmentMutex);
-		}
-	}
-#endif
-
-	if (_cacheletHead != NULL) {
-		cache = _cacheletHead;
+	if (!hasClassSegmentMutex) {
+		Trc_SHR_Assert_ShouldNotHaveLocalMutex(classSegmentMutex);
+		enterLocalMutex(currentThread, classSegmentMutex, "class segment mutex", "updateROMSegmentList");
 	} else {
-		cache = _cc;
+		Trc_SHR_Assert_ShouldHaveLocalMutex(classSegmentMutex);
 	}
+#endif
 	
 	while (cache) {
-		if (cache && cache->isStarted()) {
+		if (cache->isStarted()) {
 			updateROMSegmentListForCache(currentThread, cache);
 		}
 		cache = cache->getNext();
```

Comment:
Is this loop necessary to support layered caches? If not, remove it. If so, we only need to look at the lower level caches once since they are read-only.

Reply:
We need to call this function for lower layers only during startup. Add one parameter `bool topLayerOnly` to the signature.